### PR TITLE
store: refactor build result

### PR DESCRIPTION
**Pull request by nicks**
Hello @maiamcc,

Please review the following commits I made in branch nicks/record:

77af5d3b9aa4cdf23bc773a6cf15853cb87fe9eb (2019-10-15 17:57:16 -0400)
store: refactor build result

### CODE REVIEW

**Comment by nicks** [comment]
this data structure was annoying me, and i needed a way to record the BuildType in the update history

File: internal/store/build_result.go [hunk 1]
```diff
@@ -13,94 +13,150 @@ import (
...
-		TargetID:                 id,
+func NewDockerComposeDeployResult(id model.TargetID, containerID container.ID) DockerComposeBuildResult {
+	return DockerComposeBuildResult{
+		id:                       id,
 		DockerComposeContainerID: containerID,
 	}
 }
 
+type K8sBuildResult struct {
+	// The image target that this built.
```
Comment (maiamcc) [inline internal/store/build_result.go]: ...wait, huh? wouldn't this be a k8s target ID?

Comment (nicks) [reply internal/store/build_result.go]: yep! removed

File: internal/engine/upper.go [hunk 2]
```diff
@@ -353,10 +353,14 @@ func handleBuildCompleted(ctx context.Context, engineState *store.EngineState, c
 	if mt.Manifest.IsDC() {
 		state, _ := ms.RuntimeState.(dockercompose.State)
 
-		dcResult := cb.Result[mt.Manifest.DockerComposeTarget().ID()]
-		cid := dcResult.DockerComposeContainerID
-		if cid != "" {
-			state = state.WithContainerID(cid)
+		result := cb.Result[mt.Manifest.DockerComposeTarget().ID()]
+		dcResult, ok := result.(store.DockerComposeBuildResult)
+		cid := container.ID("")
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/upper.go]: done

File: internal/engine/image_build_and_deployer.go [hunk 3]
```diff
@@ -291,7 +291,7 @@ func (ibd *ImageBuildAndDeployer) createEntitiesToDeploy(ctx context.Context,
 		}
 
 		for _, depID := range depIDs {
-			ref := results[depID].Image
+			ref := store.ImageFromBuildResult(results[depID])
 			if ref == nil {
 				return nil, fmt.Errorf("Internal error: missing build result for dependency ID: %s", depID)
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/image_build_and_deployer.go]: done

File: internal/engine/image_build_and_deployer.go [hunk 4]
```diff
@@ -373,11 +373,13 @@ func injectImageDependencies(iTarget model.ImageTarget, iTargetMap map[model.Tar
 	}
 
 	for _, dep := range deps {
-		modified, err := ast.InjectImageDigest(iTargetMap[dep.TargetID].ConfigurationRef, dep.Image)
+		image := store.ImageFromBuildResult(dep)
```
[the remark this answers is not in the document]
Comment (nicks) [reply internal/engine/image_build_and_deployer.go]: that's a nice idea in theory, but doesn't actually work in practice. go lacks generics, and the function that produces this array needs to use the more general type

**Pull request merged**